bin/cmd.js

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env node
+
+import WikibaseRepo from './../lib/repo.js'
+import { pickLanguages, pickKeys } from './../lib/util.js'
+import { execCmd } from './../lib/exec.js'
+
+;(async () => {
+  const [source, target, ...entities] = process.argv.slice(2)
+  const sourceRepo = new WikibaseRepo(source)
+  const targetRepo = new WikibaseRepo(target, {
+    oauth: {
+      consumer_key: process.env.TARGET_WIKI_OAUTH_CONSUMER_TOKEN,
+      consumer_secret: process.env.TARGET_WIKI_OAUTH_CONSUMER_SECRET,
+      token: process.env.TARGET_WIKI_OAUTH_ACCESS_TOKEN,
+      token_secret: process.env.TARGET_WIKI_OAUTH_ACCESS_SECRET
+    }
+  })
+
+  const contentLanguages = await targetRepo.getContentLanguages()
+
+  let data = await sourceRepo.getEntities(...entities)
+  data = data
+    .map(e => pickKeys(e, 'type', 'labels', 'descriptions', 'aliases', 'datatype'))
+    .map(e => pickLanguages(e, ...contentLanguages))
+
+  await targetRepo.createEntities(...data)
+  return `Sucessfully transferred ${entities.length} entities from ${source} to ${target}.`
+})()
+  .then((result) => {
+    if (result) {
+      console.log(result)
+    }
+  })
+  .catch((err) => {
+    console.error(err)
+    process.exitCode = 1
+  })
+  .then(() => {
+    if (process.env.CALLBACK_ON_SUCCESS && process.exitCode === 0) {
```

Comment:
Running the callbacks from in here now means we don't catch SIGINT, but I guess that's ok for our use case?

Reply:
I'm not sure if this might leave us with a larger (but still small) chance of wiki's being left in a partially imported state. In any case I don't think k8s is likely to SIGINT but rather SIGTERM. If we wanted to be super belt and braces we could trap this and then CALLBACK_ON_FAILURE

Reply:
There is a cleanup Laravel job that sweeps stalled imports every hour or so.

We could leave the trap in bash as is if we wanted to, but then we'd have a bit of mish-mash still. Which is what we are trying to leave behind.

Reply:
Added some Node based trapping in https://github.com/wbstack/transferbot/pull/7/commits/214be4d9973acb7e478021fccddbcd259fe27251 that seems to work when double CRL+Cing the thing running in Docker

```
➜  transferbot git:(fr/nodeify) ✗ docker run -e TARGET_WIKI_OAUTH_CONSUMER_TOKEN="xxx" -e TARGET_WIKI_OAUTH_CONSUMER_SECRET="xxx" -e TARGET_WIKI_OAUTH_ACCESS_TOKEN="xxx" -e TARGET_WIKI_OAUTH_ACCESS_SECRET="xxx" -e CALLBACK_ON_FAILURE="echo 'failing'" --rm 4ab54cf0400349e5e0afbf06e1965046c1b7cf0f8890c505afd8b733efac3a07 https://www.wikidata.org https://updatewiki.wikibase.dev Q107819273
^CAn error occurred executing the command:
Error: Received SIGINT, program will exit before finishing.
    at process.<anonymous> (file:///pkg/lib/signals.js:5:16)
    at process.emit (node:events:519:28)
failing
^C%
```

Single Ctrl+C would have Node waiting for its event loop to become empty and would trigger both the transfer as well as the failure callback. I would think this is a problem we'd also have in bash.

Reply:
I ended up needing to do a load of reading the docs about `process` and this then inspired me to ask: "should we consider using `beforeExit` for all this stuff?" see https://nodejs.org/api/process.html#event-beforeexit

Reply:
So this is a bit ugly. When using `beforeExit`, a forceful exit (i.e. Ctrl+C) or an unhandled exception will not be considered:

> The 'beforeExit' event is not emitted for conditions causing explicit termination, such as calling [process.exit()](https://nodejs.org/api/process.html#processexitcode) or uncaught exceptions.

In theory `on('exit')` would help us, however this (as opposed to `beforeExit`) only allows for synchronous code to be run in the callback, meaning we can't shell out to the command the user is passing.

Maybe we're already at a point where this is "as good as it gets" if we want to keep this in Node.

Reply:
Maybe it's good to remind ourselves to not overcomplicate this and aim for a solution that ensures the given commands will ALWAYS run as expected.

Reply:
I implemented some cancellation mechanism in 49f77f1acf9638aba49752eb1117cb6339ead55c

This works as intended now, but I am not 100% sure we should use it as it makes the code kind of hard to follow. This is mostly because the wikibase-x librarys do not expose the fetch they are using, so we cannot use the plain `signal` but have to add some custom wrapper, when `fetch` calls can use the signal as intended. I'm not really sure if the extra complexity justifies the outcome.